securing_apps/topics/oidc/keycloak-gatekeeper.adoc

```diff
@@ -0,0 +1,494 @@
+listen: 127.0.0.1:3000
+redirection-url: http://127.0.0.1:3000
+upstream-url: http://127.0.0.1:80
+resources:
+- uri: /admin*
+  methods:
+  - GET
+  roles:
+  - client:test1
+  - client:test2
+  require-any-role: true
+  groups:
+  - admins
+  - users
+- uri: /backend*
+  roles:
+  - client:test1
+- uri: /public/*
+  white-listed: true
+- uri: /favicon
+  white-listed: true
+- uri: /css/*
+  white-listed: true
+- uri: /img/*
+  white-listed: true
+----
+
+Anything defined in a configuration file can also be configured using command line options, such as in this example.
+
+[source,bash]
+----
+bin/{generic_adapter_name} \
+    --discovery-url=https://keycloak.example.com/auth/realms/<REALM_NAME> \
+    --client-id=<CLIENT_ID> \
+    --client-secret=<SECRET> \
+    --listen=127.0.0.1:3000 \ # unix sockets format unix://path
+    --redirection-url=http://127.0.0.1:3000 \
+    --enable-refresh-tokens=true \
+    --encryption-key=AgXa7xRcoClDEU0ZDSH4X0XhL5Qy2Z2j \
+    --upstream-url=http://127.0.0.1:80 \
+    --enable-default-deny=true \
+    --resources="uri=/admin*|roles=test1,test2" \
+    --resources="uri=/backend*|roles=test1" \
+    --resources="uri=/css/*|white-listed=true" \
+    --resources="uri=/img/*|white-listed=true" \
+    --resources="uri=/public/*|white-listed=true"
+----
+
+By default the roles defined on a resource perform a logical `AND` so all roles specified must be present in the claims, this behavior can be altered by the `require-any-role` option, however, so as long as one role is present the permission is granted.
+
+==== HTTP routing
+
+By default all requests will be proxyed on to the upstream, if you wish to ensure all requests are authentication you can use this:
+
+[source,bash]
+----
+--resource=uri=/* # note, unless specified the method is assumed to be 'any|ANY'
+----
+
+The HTTP routing rules follow the guidelines from link:https://github.com/go-chi/chi#router-design[chi]. The ordering of the resources do not matter, the router will handle that for you.
+
+==== Session-only cookies
+
+By default the access and refresh cookies are session-only and disposed of on browser close; you can disable this feature using the `--enable-session-cookies` option.
+
+==== Forward-signing proxy
+
+Forward-signing provides a mechanism for authentication and authorization between services using tokens issued from the IdP. When operating in this mode the proxy will automatically acquire an access token (handling the refreshing or logins on your behalf) and tag outbound requests with a Authorization header. You can control which domains are tagged with the --forwarding-domains option. Note, this option use a **contains** comparison on domains. So, if you wanted to match all domains under *.svc.cluster.local you can use: --forwarding-domain=svc.cluster.local.
+
+At present the service performs a login using oauth client_credentials grant type, so your IdP service must support direct (username/password) logins.
```

Comment:
Maybe replace `oauth client_credentials` by `OAuth 2.0 Client Credentials` ?

Reply:
Agreed, except I think only OAuth needs to be capitalized and not "client credentials".